feat: include multiple env files and specify variables

kevinbardi opened the pull request:
Hello. I really like what you did, it's simple and really useful.

I need some changes for my project, so I forked it, and I worked on a solution.

Basically what I need is:

Be able to load environment variables from multiple files
Specify which are the variables that I need, because I do not need to expose all of them, and I think it's cleaner and safer.
I updated the unit tests, and I also added some examples in the README file.

Thanks

cardinalby commented on src/main.ts:
```diff
@@ -8,20 +8,37 @@ function processValue(name: string, value: string) {
+    }
+    
+    if (inputs.variables) {
+        const names = inputs.variables.split(',')
```
"," is a bad separator, because it's a valid character that can be used in a filename.
I suggest using "|" instead

kevinbardi replied:
Updated, and I added a custom separator if they want to use it

cardinalby replied:
Please remove the "custom separator" option and stick to "|" char.

I believe it's an extra complication for the action and the documentation (coming with support burden for me), most of the people don't need these kind of flexibity, but they need the action to be understandable and ease to use.

If one wants to use another separator (comma-separated string comes from another step), they can change the separators outside the action (replace chars as a separate step), I believe it's not a feature we should support inside the action

kevinbardi replied:
Updated 👍

cardinalby commented on tests/integration/action.test.ts:
```diff
@@ -157,4 +157,25 @@ describe('export-env-action', () => {
         expect(res.warnings).toHaveLength(0);
     });
+
+    it('should process multiple files', async () => {
```
Please add tests to check if variables from one file are extended with variables from another.
Add a note to readme about the order (can variables from 2nd file be extended by variables from the 1st?)

kevinbardi replied:
Added multiple unit tests with different scenarios 👍

cardinalby commented on src/main.ts:
```diff
@@ -8,20 +8,37 @@ function processValue(name: string, value: string) {
+}
+
+function getVars(): dotenv.DotenvParseOutput {
+    const files = inputs.envFile.split(',')
```
To unify the separators I suggest using "|" here as well

kevinbardi replied:
Updated, and I added a custom separator if they want to use it

cardinalby commented on src/inputs.ts:
```diff
@@ -5,6 +5,10 @@ export default {
         return core.getInput('envFile', {required: true})
     },
 
+    get variables(): string {
```
Please update `action.yml` and `readme.md` with a new input (and with a comment to `envFile` input as well)

kevinbardi replied:
Updated. I also added a table of content to the readme, so it's easier to find inputs, outputs, examples, etc...

cardinalby commented on src/main.ts:
```diff
@@ -4,24 +4,45 @@ import * as dotenvExpand from "dotenv-expand";
+        vars = dotenvExpand.expand({parsed: vars, ignoreProcessEnv: !inputs.expandWithJobEnv}).parsed as dotenv.DotenvParseOutput
+    }
+    
+    if (inputs.variables && inputs.variables.toLocaleLowerCase() !== 'all') {
```
a variable with the name `all` can exist so it's not correct

kevinbardi replied:
well...but every name is valid for a variable...you can also have a variable called `true="myValue"`
The intention here is...if variable param is not defined, or you use 'all', you retrieve all the variables and you will have access to them in the output. Usually you have both options to specify if you want to include all of them, but if you prefer to remove 'all', I can remove it, and the only option to generate an output with all the variables is to do not use variables definition.
This is up to you...I can remove it if you want

cardinalby replied:
In your approach if I want to export only a variable with the name `all` it will not work which is confusing.

One of the option is to call this input `filter` and:
1. if it's empty, it will allow all variables to be exported
2. if it's not empty it will work as a regexp that will allow variable names matching the regexp to be exported

It will:
1. implement the similar to file separators (`|`) but more powerful semantics (`|`  means "OR" in regexps):
    * `VAR1|VAR2`
    * ` CONFIG_.*?|PARAM_\d*`
3. keep backwards compatibility if the input not set

Share your thoughts about that

kevinbardi replied:
Updated to RegExp, and remove all as a value

cardinalby commented on action.yml:
```diff
@@ -3,8 +3,16 @@ description: 'Exports .env file variables to job env'
-    description: 'Path to .env file'
+    description: 'Path to .env file or Path to multiple .env files concatenated by `|` or custom "separator" prop'
     required: true
+  variables:
```
Introducing a new feature like that you should consider the following questions in order not to broke existing features and keep the API design consistent:

1. Do you want it to filter only exported variables or action inputs as well (in case of `export: false`)? 
2. Should it prevent not listed variables from participating in extending of the listed variables?
3. Should it impact masking values (`mask: true`)?
4. Is it possible to have a list of "excluded" variables instead?

Please describe the desired behaviour (then it should included to documentation and to test use-cases). If answers turn our too complicated it can be a reason to avoid adding this feature -> think about users who are not familiar with the details and just want to use the action in a simple case

kevinbardi replied:
Good questions. I do not want to expose to the action all the environment variables...I just need some of them, and I never know if there are new environment variables added, so excluded is not enough, because I want to retrieve only the once that I need.
So, to answer each point:

1. I want to filter in both scenarios...it doesn't matter if it's action inputs or exported variables, it makes sense to specify and avoid mistakes or wrong usage of another variable...so it's a global definition.
2. That is the tricky one...but I think it makes sense to participate on extending variables, because if you need a variable that extends from other one, you should get the right value...it doesn't matter if you don't need in the output the intermediate variable.
3. Yes, I think everything should work as it was working from before...the only thing that changes is the final output/export...but everything inside the process should work as it was working
4. From my perspective it doesn't make sense. I worked with secrets and environment variables from a security perspective many times...and the idea is always to only expose what you need, nothing else. And if you exclude some variables is not the same behaviour...because any new addition to the .env file, you have to keep updated the excluded definition, when probably that new variable is not associated neither required in the action.
5. I added unit tests to cover different workflows related to variables with extend, mask, etc...from before. I think I covered every case.
6. And I also updated the documentation and action.yml with this definition. You can find input definition in the README file for the variables, and also examples with variables

cardinalby replied:
Agree on all points, just wanted to ensure you understand the desired behaviour and to be explicit about it, but this behaviour should be documented

kevinbardi replied:
Documentation is already updated with filter and everything

kevinbardi commented:
I think I need to update github workflow action to use files in my new tests...doing it now @cardinalby

cardinalby commented:
> I think I need to update github workflow action to use files in my new tests...doing it now

Don't think so... integration tests step fails. Does it pass locally?

kevinbardi commented:
> > I think I need to update github workflow action to use files in my new tests...doing it now
> 
> Don't think so... integration tests step fails. Does it pass locally?

yes, it's working locally. I'll take a look on this later today

kevinbardi commented:
Funny, I found the issue. In the Github action we are testing over `dist/index.js` file (generated with `pack` script), but locally it is executed over lib folder. For that reason it was working locally. It's not in github action the automation to generate that dist/index.js file, but it's being used as main file.
I updated the file and now it seems to be working on my repo and local

kevinbardi commented:
@cardinalby it should be ready to approve the workflow to run, and continue with the process 😄

cardinalby commented:
> Funny, I found the issue. In the Github action we are testing over `dist/index.js` file (generated with `pack` script), but locally it is executed over lib folder. For that reason it was working locally. It's not in github action the automation to generate that dist/index.js file, but it's being used as main file. I updated the file and now it seems to be working on my repo and local

I forgot to warn you about this behavior. Actually, it's mine actions integration tests lib :)

kevinbardi commented:
yes, I saw in the repo that was also yours. I checked on the code and I found it was loading everything from dist, and I saw in this repo we are generating the dist folder with a different script, and we needed to push that. Everything fixed so. Is there any pending update? Can you merge it and generate a new version this week by any chance? So I can start using it directly from the action in my CI/CD process 😄

cardinalby commented:
Please answer on the remaining comments

kevinbardi commented:
> Please answer on the remaining comments

Sorry, which are the remaining comments? I think I answered all of them. I cannot find pending updates. I added more unit tests, I fixed the test workflow and I removed the separator variable. Sorry, but not sure what is missing...

cardinalby commented:
<img width="820" alt="image" src="https://github.com/cardinalby/export-env-action/assets/2001539/545c5b95-d0b6-4161-8ced-b84a8e2a36d8">

cardinalby commented:
Sorry, I havent' added them, they were in a pending review :(

cardinalby commented on src/main.ts:
```diff
@@ -4,24 +4,43 @@ import * as dotenvExpand from "dotenv-expand";
+    }
+    
+    if (inputs.filter) {
+        const criteria = new RegExp(inputs.filter)
```
please catch a possible exception from `new RegExp(...)` and re-throw it with a message saying that there is an error in `filter` input regexp

kevinbardi replied:
Done 👍

cardinalby commented on README.md:
````diff
@@ -70,11 +149,30 @@ VAR2=def
+VAR_2=bbb
+VAR_3=ccc
+```
+Will lead to following exported variables: `VAR_1 = aaa`, `VAR3 = ccc`.
````
Please add here information regarding the questions we discussed:

1. Do you want it to filter only exported variables or action inputs as well (in case of export: false)?
2. Should it prevent not listed variables from participating in extending of the listed variables?
2. Should it impact masking values (mask: true)?

Smth like:
```
Filters both exported variables and action outputs (if `export: false`), but doesn't impacts variables extending and masking.
```

kevinbardi replied:
Done 👍

cardinalby commented:
Great job! Thanks for contributing

kevinbardi commented:
Hello @cardinalby , could you make a release with the latest changes please?

cardinalby commented:
done. it is also available under `@2.2.0`, `@v2`, `@master`
